r/R/dplyr-functions.R

```diff
@@ -634,20 +634,45 @@ nse_funcs$wday <- function(x, label = FALSE, abbr = TRUE, week_start = getOption
 }
 
 nse_funcs$log <- function(x, base = exp(1)) {
-  
+
   if (base == exp(1)) {
     return(Expression$create("ln_checked", x))
   }
-  
+
   if (base == 2) {
     return(Expression$create("log2_checked", x))
   }
-  
+
   if (base == 10) {
     return(Expression$create("log10_checked", x))
-  } 
+  }
   # ARROW-13345
   stop("`base` values other than exp(1), 2 and 10 not supported in Arrow", call. = FALSE)
 }
 
 nse_funcs$logb <- nse_funcs$log
+
+nse_funcs$if_else <- function(condition, true, false, missing = NULL){
+  if (!is.null(missing)) {
+    arrow_not_supported("missing argument")
+  }
+
+  # TODO: if_else doesn't yet support factors/dictionaries this can be removed when
+  # ARROW-13358 merges
+  warn_r_types <- is.factor(true) || is.factor(false)
+  # However, if they are expressions, we need to use the functions from nse_funcs
+  warn_expression_types_true <- inherits(true, "Expression") &&  nse_funcs$is.factor(true)
+  warn_expression_types_false <- inherits(false, "Expression") && nse_funcs$is.factor(false)
+
+  if (warn_r_types | warn_expression_types_true | warn_expression_types_false) {
+    warning("Factors are currently converted to chracters in if_else and ifelse", call. = FALSE)
+  }
```

Comment:
I implemented the warning, but we can take this out if we think it would be better to not.

Reply:
They are converted? Where/how?

Also "chracters" is misspelled here and in the test

Reply:
The likely source of this conversion is https://github.com/apache/arrow/blob/f7cac0e5866a08e946a14976ef854e8c42ac281b/cpp/src/arrow/compute/kernels/scalar_if_else.cc#L980 (h/t @lidavidm ) When `dispatchbest` ing, dictionaries are decoded (unless there is a specific kernel that takes dictionaries).

I've corrected the misspelling

Reply:
Is this unique to if_else or a more general issue? If the latter, it might be better just documented somewhere (probably, tbh, in a vignette we haven't written yet). 

If we're keeping the warning, I'm not sure we should (only) say "factor" or "character" since it's actually about the Arrow types.

Reply:
This is a more general issue (e.g. I believe all arithmetic kernels do the same thing). However for those kernels it's more unavoidable; for if_else and friends maybe we should implement specific dictionary handling.

Reply:
e.g case_when does the same, so will coalesce when merged.

Reply:
I've added both dictionary+factor and string+character to the warning. I agree about talking about this type of casting/coalescing in a vignette somewhere when it happens in the cases where it's more-or-less unavoidable. Though in this case, like @lidavidm says, the casting is (hopefully) only temporary so we should be able to pull this out next release (or maybe even before then if the extra functionality is added soon enough).